internal/rootcoord/import_manager.go

```diff
@@ -462,181 +596,193 @@ func (m *importManager) getTaskState(tID int64) *milvuspb.GetImportStateResponse
-			resp.State = v.GetState().GetStateCode()
-			resp.RowCount = v.GetState().GetRowCount()
-			resp.IdList = v.GetState().GetRowIds()
-			resp.Infos = append(resp.Infos, &commonpb.KeyValuePair{Key: Files, Value: strings.Join(v.GetFiles(), ",")})
-			resp.Infos = append(resp.Infos, &commonpb.KeyValuePair{
-				Key:   FailedReason,
-				Value: v.GetState().GetErrorMessage(),
-			})
-			resp.DataQueryable = v.GetDataQueryable()
-			resp.DataIndexed = v.GetDataIndexed()
-			m.getCollectionPartitionName(v, resp)
-		}
-	}()
 	if found {
 		return resp
 	}
 	log.Debug("get import task state failed", zap.Int64("taskID", tID))
 	return resp
 }
 
-// loadFromTaskStore loads task info from task store when RootCoord (re)starts.
-func (m *importManager) loadFromTaskStore() error {
+// loadFromTaskStore loads task info from task store (Etcd).
+// loadFromTaskStore also adds these tasks as pending import tasks, and mark
+// other in-progress tasks as failed, when `load2Mem` is set to `true`.
+// loadFromTaskStore instead returns a list of all import tasks if `load2Mem` is set to `false`.
+func (m *importManager) loadFromTaskStore(load2Mem bool) ([]*datapb.ImportTaskInfo, error) {
 	log.Info("import manager starts loading from Etcd")
 	_, v, err := m.taskStore.LoadWithPrefix(Params.RootCoordCfg.ImportTaskSubPath)
 	if err != nil {
 		log.Error("import manager failed to load from Etcd", zap.Error(err))
-		return err
+		return nil, err
 	}
-	m.workingLock.Lock()
-	defer m.workingLock.Unlock()
-	m.pendingLock.Lock()
-	defer m.pendingLock.Unlock()
+	var taskList []*datapb.ImportTaskInfo
 	for i := range v {
 		ti := &datapb.ImportTaskInfo{}
 		if err := proto.Unmarshal([]byte(v[i]), ti); err != nil {
 			log.Error("failed to unmarshal proto", zap.String("taskInfo", v[i]), zap.Error(err))
 			// Ignore bad protos.
 			continue
 		}
-		// Put tasks back to pending or working task list, given their import states.
-		if ti.GetState().GetStateCode() == commonpb.ImportState_ImportPending {
-			log.Info("task has been reloaded as a pending task", zap.Int64("task ID", ti.GetId()))
-			m.pendingTasks = append(m.pendingTasks, ti)
+		if load2Mem {
+			// Put pending tasks back to pending task list.
+			if ti.GetState().GetStateCode() == commonpb.ImportState_ImportPending {
+				log.Info("task has been reloaded as a pending task", zap.Int64("task ID", ti.GetId()))
+				m.pendingLock.Lock()
+				m.pendingTasks = append(m.pendingTasks, ti)
+				m.pendingLock.Unlock()
+			} else {
+				// other non-failed and non-completed tasks should be marked failed, so the bad s egments
+				// can be cleaned up in `removeBadImportSegmentsLoop`.
+				if ti.GetState().GetStateCode() != commonpb.ImportState_ImportFailed &&
+					ti.GetState().GetStateCode() != commonpb.ImportState_ImportFailedAndCleaned &&
+					ti.GetState().GetStateCode() != commonpb.ImportState_ImportCompleted {
+					ti.State.StateCode = commonpb.ImportState_ImportFailed
+					if err := m.persistTaskInfo(ti); err != nil {
```

Comment:
if mark failed, should return error?

Reply:
It will not return an error as other tasks are successfully load. Also it is expected to mark non-complete tasks as failed.